crafting.py

```diff
@@ -0,0 +1,316 @@
+            propositions.add(Proposition(i,k))
+     
+    return propositions
+
+
+def recipe_to_propositions(recipe: Recipe) -> Set[Proposition]:
+    propositions: Set[Proposition] = set()
+    # G. Do something with recipe.consumes, recipe.produces, and recipe.requires.
+    # Emit, for this recipe, all the propositions entailed by the preconditions and the _minimal_ set of propositions embodied in the preconditions (i.e., don't need to output wood >= 2, wood >= 1, wood >= 0 if the recipe creates 2 wood.)
+    r = state_propositions(recipe.requires)
+    p = state_propositions(recipe.produces)
+    c = state_propositions(recipe.consumes)
+                    
+    propositions |= r     
+    propositions |= c    
+    propositions |= p   
+   
+        
+    return propositions 
+
+    
+def preconditions_satisfied(state: State, recipe: Recipe) -> bool:
+  
+    if ( state >= recipe.requires and state >= recipe.consumes) :
+        return True
+    
+    return False
+
+
+
+def apply_effects(state: State, recipe: Recipe) -> State:
+    # E. How do you make a new state out of a state and a recipe?
+    # Note, DO NOT change state in-place!
+     newstate  =  state - recipe.consumes + recipe.produces
+    
+     return newstate 
+ 
+def see_state(state:State, combinations:List[Set[Proposition]], seen_combinations:Set[FrozenSet[Proposition]]) -> bool:
+    any_new = False
+    state_props = state_propositions(state)
+    for combo in combinations:
+        if combo in seen_combinations:
+            continue 
+        
+        if(state_props.issuperset(combo)) :
+            seen_combinations.add(combo)
+            any_new = True
+        
+    return any_new
+
+   
+recipe_propositions = set()
+for r in recipes.values():
+    recipe_propositions |= recipe_to_propositions(r)    
+
+def plan_dijkstra(initial: State, goal: State, limit:int) -> Tuple[int, int, Optional[List[str]]]:
+    start_time = time.time()
+    open_list: List[Tuple[int, State,str]] = [(0, initial, "")]         
+    best_costs: Dict[State, Tuple[int, str,State]] = {initial:(0, "",initial)}
+    visit_count = 0
+    found = False
+    while open_list:
+        # Dijkstra's search uses the priority queue data structure
+        cost, curstate , gotfrom  = heapq.heappop(open_list)       
+        #print("cost=",cost,"curstate=",curstate,"from=",gotfrom)
+       
+        if(curstate >= goal):
+            x,rname,z = best_costs.get(curstate)
+            print ("found it!! Number of visited nodes=" ,visit_count , "state=",curstate,"goal=",goal,"by=",rname) 
+            found = True;
+            break;
+                
+        visit_count += 1
+        if(visit_count > limit) :
+            print("too many iterations , current state=",curstate)            
+            break
+        
+                         
+        for candidaterecipename in recipes:
+            if (preconditions_satisfied(curstate,recipes[candidaterecipename])) == False:
```

Comment:
I feel like using the `not` operator is cleaner than `== False`.

Reply:
Or not checking that it is false and then nesting all the subsequent code in this block.